contracts/contracts/SNStateProofVerifier.sol

```diff
@@ -123,6 +124,27 @@ contract SNStateProofVerifier is
         return hashvalue;
     }
 
+    // based on https://docs.starknet.io/documentation/architecture_and_concepts/Hashing/hash-functions/
+    function poseidonHashMany(
+        uint256[] memory elems
+    ) public view returns (uint256) {
+        uint256[3] memory state = [uint256(0), uint256(0), uint256(0)];
+
+        for (uint256 i = 0; i < elems.length / 2; i++) {
+            state[0] = (state[0] + elems[2 * i]) % BIG_PRIME;
+            state[1] = (state[1] + elems[2 * i + 1]) % BIG_PRIME;
+            state = poseidon.poseidon(state);
+        }
```

Comment:
alternative to avoid division and multiplication:

```solidity
for (uint256 i = 0; i < elems.length; i += 2) {
    state[0] = (state[0] + elems[i]) % BIG_PRIME;
    state[1] = (state[1] + elems[i + 1]) % BIG_PRIME;
    state = poseidon.poseidon(state);
}
```

Reply:
but with (i < elems.length -1) but yes thanks for the suggestion.